README.md
# Design Evolver

Design Evolver is an optimizer that runs on top of the [j5](https://j5.jbei.org/) DNA assembly design software.  j5 (without Design Evolver) does not consider many of the possible ways to design a DNA assembly. Design Evolver gives j5 more degrees of freedom to find assemblies that are more likely to assemble successfully.  This workflow is particularly useful for creating difficult to assemble DNA such as, sequences with repeats or DNA coding for RNA with strong secondary structures.

## Depedencies

* [Python](https://www.python.org/) v3.3.x
* [Numpy](https://github.com/numpy/numpy) v1.7.1
* [DEAP](https://github.com/DEAP/deap/) v0.91
* [BioPython](https://github.com/biopython/biopython) v1.62
* [j5](https://j5.jbei.org/) v2.2.5 (not required if using a remote j5 server)

Newer versions of these dependencies may work but have not yet been tested by the maintainers.

## Installation:
1. If you plan to run j5 locally, then follow the installation instructions for [j5](https://j5.jbei.org/).
1. Clone this git repository to your local machine.
1. In the directory containing this README.md file, run: `pip3 install -r requirements.txt`

## Running Design Evolver
1. See example invocation in the [example_usage](/example_usage) file.
   * You must edit this file before using it as you must supply a j5 username and password.
1. See the [j5 manual](https://j5usermanual.lbl.gov/) for definitions of the input and output files.

## Limitations
Design Evolver does not support all j5 inputs. The following currently cannot be optimized by Design Evolver:
1. Combinatorial designs

## Bug reports and comments:
Please use the [associated issue tracker](https://github.com/wholtz/designEvolver/issues) on GitHub.

## Commercial Licensing

As noted in the [LICENSE](/LICENSE) file, commercial usage is not permitted without an additional license. To obtain a license for commercial usage contact the [University of California Tech Transfer Office](https://techtransfer.universityofcalifornia.edu/NCD/23597.html).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2013-09-27

### Added

- Initial release, all code is new and by [@wholtz](https://github.com/wholtz/).



designEvolver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright ©2013. The Regents of the University of California (Regents). All
# Rights Reserved. Permission to use, copy, modify, and distribute this
# software and its documentation for educational, research, and
# not-for-profit purposes, without fee and without a signed licensing
# agreement, is hereby granted, provided that the above copyright notice,
# this paragraph and the following two paragraphs appear in all copies,
# modifications, and distributions. Contact The Office of Technology
# Licensing, UC Berkeley, 2150 Shattuck Avenue, Suite 510, Berkeley, CA
# 94720-1620, (510) 643-7201, for commercial licensing opportunities.
# 
# Created by William J. Holtz, The California Institute for Quantitative
# Biosciences, University of California, Berkeley.
# 
# IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
# SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
# ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
# REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# 
# REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE.  THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED
# HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE
# MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.

import tempfile
import shutil
import re
import sys
import j5lib
import tweaks
import argparse
import copy
import random
import xmlrpc.client as xr
from deap import algorithms
from deap import base
from deap import creator
from deap import tools

random.seed(123456789)  # provide a seed to keep runs repeatable

parser = argparse.ArgumentParser(description='submit design to j5')
parser.add_argument('--j5URL', type=str,
                    default= 'http://j5.jbei.org/bin/j5_xml_rpc.pl',
                    help='Location of j5 XMLRPC server interface')
parser.add_argument('--username', type=str, required=True,
                    help='Username for logging into the j5 server')
parser.add_argument('--password', type=str, required=True,
                    help='password for logging into the j5 server')
parser.add_argument('--param', type=str, default='j5parameters.csv',
                    help='filename for j5 parameters')
parser.add_argument('--seq', type=str, default='mastersequences.csv',
                    help='filename for sequence name mapping')
parser.add_argument('--zipseq', type=str, default='masterzippedsequences.zip',
                    help='filename for zipped sequence files')
parser.add_argument('--parts', type=str,
                    default='partslist.csv', help='filename for parts list file')
parser.add_argument('--order', type=str, default='targetpartsorder.csv',
                    help='filename for target part order file')
parser.add_argument('--eugene', type=str,
                    default='eugeneruleslist.eug', help='filename for eugene rules')
parser.add_argument('--plasmid', type=str,
                    default='j5_plasmids.csv', help='filename for plasmid list')
parser.add_argument('--oligo', type=str,
                    default='j5_oligos.csv', help='filename for oligo list')
parser.add_argument('--direct', type=str, default='j5_directsyntheses.csv',
                    help='filename for direct syntheses list')
parser.add_argument('--diff', type=str, default='',
                    help='genbank filename to compare final sequence against')
parser.add_argument('--outdir', type=str, default='./DE/',
                    help='directory to write output and temporary files to')
parser.add_argument('--numgen', type=int, default=10,
                    help='number of generations')
parser.add_argument('--maxpop', type=int, default=50,
                    help='maximum population size')
parser.add_argument('--tweaks', type=int, default=30,
                    help='number of tweaks per design')
parser.add_argument('--multiplier', type=float, default=4,
                    help='childern per parent')
parser.add_argument('--serial', dest='serial', action='store_true',
                    help='do not evalute designs in parallel')
parser.set_defaults(serial=False)
parser.add_argument('--nosubmit', dest='nosubmit', action='store_true',
                    help='checks that design can be loaded, but not sent to j5')
parser.set_defaults(nosubmit=False)
parser.add_argument('--verbose', dest='verbose', action='store_true',
                    help='output more information')
parser.set_defaults(verbose=False)
args = parser.parse_args()

maxPopulation = args.maxpop
mu = args.maxpop
lambda_ = int(args.maxpop*args.multiplier)
maxProcesses = min(40, mu+lambda_)
initialTweaksPerList = args.tweaks
generations = args.numgen
mutationRate = 0.5
mutationDegree = 0.3
crossRate = 0.5
hallOfFameLength = 1

inputFiles = { 'param': args.param, 'seq': args.seq, 'zipseq': args.zipseq, 'parts': args.parts, 'order': args.order, 'eugene': args.eugene}

if not args.serial:
    import multiprocessing
    import logging
    if args.verbose:
        logger = multiprocessing.log_to_stderr()
        logger.setLevel(multiprocessing.SUBDEBUG)

def removeDirectories(removeList, exceptionList = []):
    for d in set(removeList) - set(exceptionList):
        shutil.rmtree(d, ignore_errors=True)

def applyTweaks(server, design, inputFiles, individual, retry=0):
    results = None
    initialLengthPlasmid = len(design.getFinalSequence())
    tweakDesign = copy.deepcopy(design)
    individual.sortBySilence() # move non-silent tweaks to start of list
    tweakDesign.breakAllBigParts() # parts could change size due to tweaks, might create violations of max part size
    for t in individual.tweaks:
        t.execute(tweakDesign)
    tweakDesign.breakAllBigParts() # this is a poor way to ensure no big parts were generated...
    positions = tweakDesign.getMaxPosition()
    lengthPlasmid = len(tweakDesign.getFinalSequence())
    if not args.nosubmit:
        try:
            inputDir = tempfile.mkdtemp(dir=args.outdir)
            returnFromServer = server.submitToj5(tweakDesign, inputFiles,
                    inputDir=inputDir, delInputDir=True)
        except xr.Fault as err:
            if args.verbose:
                print('caught xmlrpc.client.Fault on job submission')
                print('submit dir: '+inputDir)
                print(individual)
                print(tweakDesign.parts)
                sys.stdout.flush()
            retryPattern = re.compile("Can't extract zipped sequence file")
            if retryPattern.match(err.faultString) and retry<3:
                if args.verbose:
                    print('caught ioclt fault on job submission, retrying...')
                    sys.stdout.flush()
                return applyTweaks(server, design, inputFiles, individual, retry=retry+1)
            warnings = 8888888
        else:
            results = j5lib.J5Results(returnFromServer, individual.outputDir)
            warnings = results.warnings[0]
    else:
        warnings = 9999999
    return (positions, lengthPlasmid, warnings, len(individual.tweaks))

design = j5lib.J5Design()
design.loadFromFiles(
    args.param, args.seq, args.zipseq, args.parts, args.order, args.eugene,
    args.plasmid, args.oligo, args.direct)
server = j5lib.J5Connection(args.username, args.password, args.j5URL)

if args.diff != '':
    try:
        s = j5lib.readGenbankSequence(args.diff)
    except TypeError:
        print("Error loading genbank file for diff")
        raise
    finalSeq = list(s.values())[0]
else:
    finalSeq = design.getFinalSequence()
startPlasmidCompiled = re.compile(finalSeq.lower())
startPlasmidLength = len(finalSeq)

def isSamePlasmid(p):
    # use global for compiled re, so that it is only calculated once
    p=p.lower()
    return (re.search(startPlasmidCompiled, p+p) != None) and (len(p) == startPlasmidLength)


# three weights are
# (maxPositionNumber, lengthPlasmid, number of warnings, number of tweaks)
creator.create("FitnessMulti", base.Fitness, weights=(-10000, -1, -1000000, -10))
creator.create("Individual", j5evo.J5TweakSet, fitness=creator.FitnessMulti)
toolbox = base.Toolbox()
toolbox.register("individual", creator.Individual, args.tweaks, args.outdir)
toolbox.register("population", tools.initRepeat, list, toolbox.individual)
toolbox.register("evaluate", applyTweaks, server, design, inputFiles)
toolbox.register("mate", j5evo.mateJ5TweakSets)
toolbox.register("mutate", j5evo.mutateTweakSet, mutationDegree)
toolbox.register("select", tools.selTournament, tournsize=maxPopulation)

if __name__ == '__main__':
    if (args.verbose):
        print(args)
    if not args.serial:
        pool = multiprocessing.Pool(processes=maxProcesses)
        toolbox.register("map", pool.map)
    tempDirectories = []
    pop = toolbox.population(n=lambda_+mu)
    best = tools.HallOfFame(hallOfFameLength)
    history = tools.History()
    toolbox.decorate("mate", history.decorator)
    toolbox.decorate("mutate", history.decorator)
    history.update(pop)
    stats = tools.Statistics(lambda ind: ind.fitness.values)
    stats.register("avg", tools.mean)
    stats.register("std", tools.std)
    stats.register("min", min)
    stats.register("max", max)

    for g in range(0,generations):
        pop = j5evo.uniqueRepr(pop) # remove duplicates
        while len(pop) < 2:
            pop.append(toolbox.individual())
        algorithms.eaMuPlusLambda(pop, toolbox, mu=mu, lambda_=lambda_,
                        cxpb=crossRate, mutpb=mutationRate,
                        ngen=1, stats=stats, halloffame=best,
                        verbose=args.verbose)
        tempDirectories.extend([i.outputDir for i in history.genealogy_history.values()])
        removeDirectories(tempDirectories, [i.outputDir for i in best])
    print('output from best design: '+str(best[0].outputDir))



tweaks.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright ©2013. The Regents of the University of California (Regents). All
# Rights Reserved. Permission to use, copy, modify, and distribute this
# software and its documentation for educational, research, and
# not-for-profit purposes, without fee and without a signed licensing
# agreement, is hereby granted, provided that the above copyright notice,
# this paragraph and the following two paragraphs appear in all copies,
# modifications, and distributions. Contact The Office of Technology
# Licensing, UC Berkeley, 2150 Shattuck Avenue, Suite 510, Berkeley, CA
# 94720-1620, (510) 643-7201, for commercial licensing opportunities.
# 
# Created by William J. Holtz, The California Institute for Quantitative
# Biosciences, University of California, Berkeley.
# 
# IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
# SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
# ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
# REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# 
# REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE.  THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED
# HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE
# MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.

import j5lib
import numpy
import tempfile
import random
import math
import copy

random.seed(123456789)  # provide a seed to keep runs repeatable


class J5Tweak(object):

    def __init__(self, parameters = None):
        if parameters is None:
            self.parameters = {}
            self.mutate(1)  # initialize parameters to random values
        else:
            self.parameters = parameters

    def __repr__(self):
        d = dictToOrderedString(self.parameters)
        return "j5evo."+self.__class__.__name__ + "(" + d +")"

    def __cmp__(self, other):
        if self.__repr__ > other.__repr__:
            return 1
        if self.__repr__ < other.__repr__:
            return -1
        return 0

    def execute(self, design):
        raise j5Error('J5Tweak.execute()  was not overriden by child class:'+self.__class__.__name__)

    def mutate(self, degree=1):
        # degree should be a float in range [0,1] where 0 is no mutation and
        # 1 is maximal mutation(aka random parameters)
        raise j5Error('J5Tweak.mutate() was not overriden by child class:'+self.__class__.__name__)

class J5TweakSet(object):

    def __init__(self, numTweaks, baseOutputDir):
        self.baseOutputDir = baseOutputDir
        self.updateTweaks([generateRandomTweak() for i in range(0, numTweaks)])

    def __repr__(self):
        # return self.outputDir +' '+str(self.tweaks)
        return str(self.tweaks)

    def __cmp__(self, other):
        if self.tweaks > other.tweaks:
            return 1
        if  self.tweaks < other.tweaks:
            return -1
        return 0


    def setNewOutputDir(self):
        self.outputDir = tempfile.mkdtemp(dir=self.baseOutputDir)

    def updateTweaks(self, tweakList):
        self.tweaks = tweakList
        self.setNewOutputDir()

    def getRandomHeadOfTweakList(self):
        # return between 1 and all head members of list
        return [self.tweaks[i] for i in range(0, random.choice(range(0, len(self.tweaks))))]

    def getRandomTailOfTweakList(self):
        # return between 1 and all tail members of list
        return [self.tweaks[i] for i in range(random.choice(range(0, len(self.tweaks))), len(self.tweaks))]

    def mutate(self, mutationDegree):
        if random.random()< mutationDegree:
           del self.tweaks[random.choice(range(0,len(self.tweaks)))] 
        for t in self.tweaks:
            t.mutate(mutationDegree)
        if random.random()< mutationDegree:
            pos = random.choice(range(0, len(self.tweaks)+1))
            self.tweaks.insert(pos, generateRandomTweak())
        if random.random()< mutationDegree:
            # swap positions of two tweaks
            r1 = random.choice(range(0,len(self.tweaks))) # random index from self.tweaks
            r2 = random.choice(range(0,len(self.tweaks))) # random index from self.tweaks
            self.tweaks[r2], self.tweaks[r1] = self.tweaks[r1], self.tweaks[r2]
        if len(self.tweaks) == 0: # don't let there be empty tweak lists
            self.tweaks.append(generateRandomTweak())
        self.updateTweaks(self.tweaks)


    def sortBySilence(self):
        # move all non-silent tweaks to the start of the list
        # do not change the relative order within the non-silent 
        # tweaks or within the silent tweaks
        nonSilent = []
        silent = []
        nonSilentTweaks = ['ModifyPartEnd']
        for t in self.tweaks:
            if t.__class__.__name__ in nonSilentTweaks:
                nonSilent.append(t)
            else:
                silent.append(t)
        self.tweaks = nonSilent + silent


def sortedOnRepr(l):
    return sorted(l, key=repr)


def uniqueRepr(l):
    if len(l) <= 1:
        return l
    s = sortedOnRepr(l)
    same = [a==b for (a, b) in zip(s, s[1:]+[s[0]])]
    if sum(same) == len(same): # if all equal
        return [s[0]]
    for i in reversed(range(0,len(s))):
        if same[i]:
            del s[i]
    return s


def dictToOrderedString(d):
    out = '{'
    firstLoop = True
    for k in sorted(d.keys()):
        if not firstLoop:
            out = out + ', '
        else:
            firstLoop = False
        out = out + repr(k) + ': '+repr(d[k])
    return out + '}'


def generateRandomTweak():
    possibleTweaks = [SilentJunctionShift, ReducePartsSilently, SplitPart, ForcedRelativeOverhang, Extra5pCPECOverlap, Extra3pCPECOverlap, ModifyPartEnd]
    return random.choice(possibleTweaks)() # return an instance


def mutateTweakSet(mutationDegree, tweakSet):
    tweakSetCopy = copy.deepcopy(tweakSet)
    tweakSetCopy.mutate(mutationDegree)
    return tweakSetCopy,

def areEqualDicts(a,b):
    if set(a.keys()) != set(b.keys()):
        return False
    for k in a.keys():
        if a[k] != b[k]:
            return False
    return True

def areLessThanDicts(a,b):
    # fairly arbitrary way of comparing dictionaries,
    # but more robust than saying equal key set and equal value sets
    # implies equal dictionaries (wrong, because values could be swapped)
    ka = sorted(set(a.keys))
    kb = sorted(set(b.keys))
    if  ka < kb:
        return True
    if ka > kb:
        return False
    for k in ka:
        if a[k] < b[k]:
            return True
    return False

def areEqualTweaks(a,b):
    if a.__class__.__name__ != b.__class__.__name__:
        return False
    return areEqualDicts(a.parameters, b.parameters)

def areLessThanTweaks(a,b):
    if a.__class__.__name__ < b.__class__.__name__:
        return True
    if a.__class__.__name__ > b.__class__.__name__:
        return False
    return areLessThanDicts(a.parameters, b.parameters)

def areEqualTweakSets(a,b):
    if len(a.tweaks) != len(b.tweaks):
        return False
    for (ta, tb) in  zip(A.tweaks, B.tweaks):
        if not areEqualTweaks(ta,tb):
            return False
    return True

def areLessThanTweakSets(a,b):
    if len(a.tweaks) < len(b.tweaks):
        return True
    if len(a.tweaks) > len(b.tweaks):
        return False
    for (ta, tb) in zip(a.tweaks, b.tweaks):
        if areLessThanTweaks(ta,tb):
            return True
    return False

def mateJ5TweakSets(ts1, ts2):
    #copy J5TweakSets before calling this function. They are modified in place
    ts1.updateTweaks(ts1.getRandomHeadOfTweakList()+ts2.getRandomTailOfTweakList())
    ts2.updateTweaks(ts2.getRandomHeadOfTweakList()+ts1.getRandomTailOfTweakList())
    return ts1, ts2


class SinglePartSingleParameter(J5Tweak):

    def mutate(self, degree=1, maxVariance=6, canBeNegative = False):
        # variance of geometic distribution = (1-p)/p^2
        v = float(degree)*float(maxVariance)
        p=(((4*v+1)**0.5)-1)/(2*v)
        if canBeNegative:
            posNeg = random.choice([-1, 1])
        else:
            posNeg = 1
        randomAmount = numpy.random.geometric(p) * posNeg
        if degree == 1:
            self.parameters['position'] = int(
                math.floor(random.random() * 10000))  # this will get wrapped via %
            self.parameters['amount'] = randomAmount
        else:
            if random.random() < degree:
                self.parameters['position'] += random.choice([-1, 1])
            self.parameters['amount'] += randomAmount


class SinglePartVectorParameter(SinglePartSingleParameter):
    # add an end of the DNA strand to SinglePartSingleParameter
    def mutate(self, degree=1, maxVariance=6, canBeNegative=False):
        super(SinglePartVectorParameter, self).mutate(degree, maxVariance, canBeNegative)
        if random.random() < degree:
            self.parameters['end'] = random.choice(['end3p', 'end5p'])


class EndShift(J5Tweak): # for inheritance only. Not to be directly instantiated

    def mutate(self, degree=1):
        if degree == 1:
            self.parameters['fractionOfMaxShift'] = random.random()
            self.parameters['position'] = int(
                math.floor(random.random() * 10000))  # this will get wrapped via %
        if random.random() < degree:
            self.parameters['end'] = random.choice(['end3p', 'end5p'])
        if random.random() < degree:
            self.parameters['position'] += random.choice([-1, 1])
        self.parameters['fractionOfMaxShift'] = random.random() * degree + self.parameters[
            'fractionOfMaxShift'] * (1 - degree)


class ModifyPartEnd(SinglePartVectorParameter):

    def mutate(self, degree=1, maxVariance=20, canBeNegative=True):
        super(ModifyPartEnd, self).mutate(degree, maxVariance, canBeNegative)

    def execute(self, design):
        # end3p and end5p refer to final assembed orientation of part
        part = design.getPartByPosition(self.parameters['position'])
        if self.parameters['end'] != 'end3p' and self.parameters['end'] != 'end5p':
            raise j5Error( "ModifyPartEnd.parameters['end'] != 'end3p' or 'end5p'")
        modifyStart = part.useInSourceDirection() == (self.parameters['end'] == 'end5p')
        part.updateEdgeAndCheckBounds(design, self.parameters['amount'], modifyStart)


class SilentJunctionShift(EndShift):

    def execute(self, design):
        shifts = design.findSilentJunctionShifts()
        part = design.getPartByPosition(self.parameters['position'])
        if self.parameters['end'] == 'end3p':
            direction = 1
            neighborPart = design.getNextPart(part)
        elif self.parameters['end'] == 'end5p':
            direction = -1
            neighborPart = design.getPreviousPart(part)
        else:
            raise j5Error( "SilentJunctionShift.parameters['end'] != 'end3p' or 'end5p'")
        maxShiftDistance = min(shifts[part.position][1+direction],len(design.getPartSequence(neighborPart))-1)
        if self.parameters['end'] == 'end5p': # need to perform shift from part 5 prime of the junction
            part = neighborPart
        distance = direction*int(
            math.ceil(maxShiftDistance * self.parameters['fractionOfMaxShift']))
        design.shiftPartJunction(part, distance)


class ReducePartsSilently(J5Tweak):

    def execute(self, design):
        design.reducePartsSilently()

    def mutate(self, degree=1):
        pass # there are no parameters to mutate


class SplitPart(J5Tweak):

    def execute(self, design):
        position = design.resolveWrappedAssemblyPosition(self.parameters['position'])
        design.breakPart(position)

    def mutate(self, degree=1):
        if degree == 1:
            self.parameters['position'] = int(
                math.floor(random.random() * 10000))  # this will get wrapped via %
        if random.random() < degree:
            self.parameters['position'] += random.choice([-1, 1])


class ForcedRelativeOverhang(SinglePartSingleParameter):

    def execute(self, design):
        part = design.getPartByPosition(self.parameters['position'])
        part.forcedRelativeOverhang = self.parameters['amount']
    
    def mutate(self, degree=1):
        super(ForcedRelativeOverhang, self).mutate(degree, maxVariance = 8, canBeNegative = True) 

class Extra5pCPECOverlap(SinglePartSingleParameter):

    def execute(self, design):
        part = design.getPartByPosition(self.parameters['position'])
        part.extra5pCPECOverlap= self.parameters['amount']

    def mutate(self, degree=1):
        super(Extra5pCPECOverlap, self).mutate(degree, maxVariance = 5, canBeNegative = False) 


class Extra3pCPECOverlap(SinglePartSingleParameter):

    def execute(self, design):
        part = design.getPartByPosition(self.parameters['position'])
        part.extra3pCPECOverlap= self.parameters['amount']

    def mutate(self, degree=1):
        super(Extra3pCPECOverlap, self).mutate(degree, maxVariance = 5, canBeNegative = False)



if __name__ == "__main__":
    print("Python library for evolving j5 designs")

